### ISSUE TITLE
Cannot parse vue file when it uses dynamic import syntax

### ISSUE THREAD

**Issue by bengallienne** [vue-styleguidist/vue-docgen-api#25]
Hi,

This is a great tool for documenting stuff. However when we try and use it against a vue file that contains dynamic import syntax, like this;

'mydynamic-component' : () => import(/* webpackChunkName: "mydynamic-component" */ '../mydynamic-component');

The parser throws an error **SyntaxError: unknown: Unexpected token**

And points to the line number of the above code. 

This is dynamic import syntax to lazy load certain components only when needed. We use a .babelrc file at the root which contains

{
    "plugins": ["syntax-dynamic-import"]
}

and some additional references in the webpack.config.js file to achieve this. 

Is there a way of making this component work with files containing this kind of syntax? I don't seem to able to make that section of the file ignored by the parser, and I can't find a way around this anywhere else currently. Any help would be appreciated!

**Comment by rafaesc** [comment]
Hi @bengallienne 

It's related to this case.
https://github.com/vue-styleguidist/vue-docgen-api/issues/23

vue-docgen-api don't use webpack. but I think that it is possible to fix it.
To understand this issue, can you copy the complete script?

Regards

**Comment by bengallienne** [comment]
Sure, when you say the complete script, what do you want to see? The component declaration?

**Comment by rafaesc** [comment]
haha my english isn't the best,

I was trying to say the part of `<script></script>` of your vue component

**Comment by bengallienne** [comment]
Ah ok, it's massive unfortunately, but I can show you the specific bit causing the problem. It looks like this. I know it is this part because if I take the dynamic importing out it works as normal.

`components: {
	'ag-grid-vue' : AgGridVue,
	'ag-gridform' : () => import(/* webpackChunkName: "ag-gridform/ag-gridform" */ '../ag-gridform/ag-gridform')
},`

Does that help? It's the second import that is causing the issue

**Comment by rafaesc** [comment]
Ok, to understand, so an example would be like this

```vue
<template>
</template>

<script>
export default {
  name: "example",
  components: {
    "ag-grid-vue": AgGridVue,
    "ag-gridform": () =>
      import(/* webpackChunkName: "ag-gridform/ag-gridform" */ "../ag-gridform/ag-gridform")
  },
  mounted() {
    //...
  }
};
</script>
```

**Comment by bengallienne** [comment]
Yes that's exactly right. The reason is for performance. This way the JavaScript for the ag-gridform file only gets loaded when it is first accessed.

**Comment by rafaesc** [comment]
Ok, it should be possible to fix it, I will comment it here.

**Comment by bengallienne** [comment]
Hi,

Sorry to chase, but did you manage to get anywhere with this in the end?

Thanks!

**Comment by DanielPe05** [comment]
Any updates on this?

**Comment by bengallienne** [comment]
Any updates on this one please?

**Comment by rafaesc** [comment]
@bengallienne @DanielPe05  yes,

I reviewed how I could handle this feature, however, the problem is when it tries to parse with `jsdoc-api`.

I am working in a new version to avoid usign `jsdoc-api` through AST handling (like `react-docgen`), but it's so complex to develop to not promise that it will be release in this week, it's probable take me three or four weeks to release a beta version, but when it will be done, `vue-docgen-api` will be faster.

**Comment by Teknologica** [comment]
Considering [Async Components](https://vuejs.org/v2/guide/components-dynamic-async.html#Async-Components) are a _de facto_ Vue 2 feature I think this improvement is necessary going forward with `vue-styleguidist`. Is this still on the roadmap?

**Comment by rafaesc** [comment]
@Teknologica Yes, if you use async components instead of dynamic import, vue-docgen-api will work, because the issue is in dynamic import.

Or did you mean adding async components inside of `vue-styleguidist`?

**Comment by Teknologica** [comment]
@rafaesc I guess my post was confusing. Basically I have the same issue as @bengallienne describes.

Where this is triggering an error:
```js
components: {
      myAsyncView: () => import('some/path'),
},
```
I also had to use [Yarn's selective dependency version resolutions](https://yarnpkg.com/lang/en/docs/selective-version-resolutions/) to target `2.3.12-beta` and fix the Babel problem described in #41.

**Comment by cristijora** [comment]
Indeed this is a big problem. This is holding me back to lazy load (async import) some components which might reduce the bundle size of my app and because vue doc gen api doesn't support this, it becomes a big limitation.

**Comment by elevatebart** [comment]
Could we trick babel to transform import() into require() at runtime?

**Comment by therealpecus** [comment]
gentle push because this is indeed preventing adoption of vue-styleguidist on project optimizing for performance or having dynamic components

**Comment by therealpecus** [comment]
> Could we use babel to transform import() into require() at runtime?
> 
> @rafaesc have you tried adding a line to `evalComponentCode.js`:
> 
> ```js
> ...
> const sandbox = {
>       exports: {},
>       module: {
>         exports: {},
>       },
>       require: requireSanbox,
>       import: requireSandbox, //< ======== this line
>       document: {},
>       window: {
>         location: {},
>       },
> ...
> ```

well, I patched `getParseBabel.js` to load `@babel/plugin-syntax-dynamic-import` and `dynamic-import-node`, and dynamic imports get parsed corretly. but this only solves the script compilation error.
converting dynamic imports to deferred requires throws a different vue-docgen error, so we're back to square one. note that this errors (_Requires using expressions are not supported in examples_) makes sense, since in order to execute code blocks inside markdown, styleguidist needs to set up a vuejs execution context on its own, providing every require without being able to rely on webpack to handle dependencies.

@rafaesc since the syntax issue can be easily solved without needing to rewrite vue-docgen to use  AST parsing, could you use a webpack loader to handle execution of code samples (separating code from templates and comments)? webpack does resolve dynamic imports but I don't know how vue-styleguidist handles this part.

**Comment by rafaesc** [comment]
Hi @therealpecus, yes I thought that, but weback returns a bundle, and vue-docgen-api should return a JSON that it will be used for vue-styleguidist, in essential, vue-styleguidist generates the bundle using webpack.

https://github.com/vue-styleguidist/vue-docgen-api/issues/33#issuecomment-390333082

**Comment by smoglica** [comment]
This will solve the issue:

```
npm install babel-plugin-dynamic-import-node --save-dev
```
Which will transpile `import()` to a deferred `require()`.

**.babelrc**

```
{
  "plugins": ["syntax-dynamic-import", "dynamic-import-node"]
}
```
---

Probably It's better to use this plugin only when running the styleguide server, for example:

**package.json**

```
{
 "scripts": {
   "styleguide": "cross-env BABEL_ENV=styleguidist vue-styleguidist server",
 }
}
```
**.babelrc**

```
{
  "env": {
    "styleguidist": {
      "plugins": ["dynamic-import-node"]
     }
  }
}
```
I hope this helps.

**Comment by bengallienne** [comment]
Awesome! This totally works! Thank you so so much!

**Comment by therealpecus** [comment]
Not so fast!
Dynamic imports work once babel transforms them to requires, but only for static references. If you have a dynamic reference they will not:
```js
components: {
      myAsyncView: () => import(`some/path/${someComponent}`),
},
```
we might want to split this issue to cover this additional case
cc: @bengallienne @rafaesc

**Comment by bengallienne** [comment]
So close! We aren't using that pattern in this app so it does not affect it, but probably worth splitting out or reopening definitely. Do you have a preference @rafaesc?

Thanks for the heads up @therealpecus!

**Comment by rafaesc** [comment]
@smoglica awesome job!, thanks for the contribution.

@bengallienne yes how I said, there is a good alternative to implement it with ASC, the effort is a lot, and unfortunatly I dont have to time to do it, at least for the moment.

I would like to maintain this issue open, because it can help to another guy that has this problem.

**Comment by elevatebart** [comment]
Hello @bengallienne ! Can you please install 3.1.0 or vue-styleguidist 3 and tell me if you still have the issue ?

**Comment by bengallienne** [comment]
Hi @elevatebart ! I have installed 3.1.0 now, and removed the suggested plugin above and babel env values from the script, and all is working well! Also really fast. Great work, thanks for all your help!
